Deserialization errors in Exporter are retried to aggressively

### <!-- Component -->

<!-- Not sure- -->

### <!-- Affected version -->

<!-- -8.3 -->, <!-- -8.4 -->, <!-- -8.5 -->, <!-- -8.6 -->, <!-- -8.7 -->, <!-- -8.8 -->

### <!-- Severity -->

<!-- Low- -->

### <!-- Likelihood -->

<!-- Low_ -->

### Description

When the `ExporterDirector` fails to deserialize a record, it immediately retries without backoff. For many errors, this does not make sense at all and we should, at most, retry with exponential backoff.

### Steps to reproduce

In this specific case, the errors were triggered by an upgrade of snapshot versions where we introduced backwards incompatible changes.

### Current behavior

The director keeps retrying, each time running into an exception such as:
```
 java.lang.RuntimeException: Could not deserialize object [RoleRecord]. Deserialization stuck at offset 63 of length 63
	at io.camunda.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:48) ~[zeebe-msgpack-value-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.logstreams.impl.log.LoggedEventImpl.readValue(LoggedEventImpl.java:106) ~[zeebe-logstreams-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.stream.impl.records.RecordValues.readRecordValue(RecordValues.java:34) ~[zeebe-stream-platform-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.broker.exporter.stream.RecordExporter.wrap(RecordExporter.java:46) ~[zeebe-broker-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.broker.exporter.stream.ExporterDirector.lambda$exportEvent$19(ExporterDirector.java:642) ~[zeebe-broker-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.retry.ActorRetryMechanism.run(ActorRetryMechanism.java:28) ~[zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.retry.EndlessRetryStrategy.run(EndlessRetryStrategy.java:51) ~[zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:84) [zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:43) [zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:125) [zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:127) [zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:104) [zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:224) [zeebe-scheduler-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
Caused by: java.lang.RuntimeException: Property 'roleId' has no valid value
	at io.camunda.zeebe.msgpack.value.ObjectValue.read(ObjectValue.java:131) ~[zeebe-msgpack-value-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
	at io.camunda.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:42) ~[zeebe-msgpack-value-8.8.0-SNAPSHOT.jar:8.8.0-SNAPSHOT]
```

### Expected behavior

- No immediate retries without backoff
- No retries of errors during deserialization

### Environment

SaaS & SM

### Version

- Snapshot

### Rootcause

- No backoff used
- No error handling for errors during deserialization

### Solution ideas

- Exponential backoff for retries
- Handle deserialization errors by not retrying

### Dev -> QA handover

- Resources:
- Versions to validate:
- Release version (in which version this feature will be released):


### Links

[Error Report](https://console.cloud.google.com/errors/detail/COHDsen37-GgUg;service=zeebe;time=P7D;locations=global?inv=1&invt=AbxL3g&project=camunda-saas-int)

--- a/zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirector.java
+++ b/zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirector.java
@@ -27,9 +27,9 @@ import io.camunda.zeebe.scheduler.SchedulingHints;
 import io.camunda.zeebe.scheduler.future.ActorFuture;
 import io.camunda.zeebe.scheduler.future.CompletableActorFuture;
 import io.camunda.zeebe.scheduler.retry.BackOffRetryStrategy;
-import io.camunda.zeebe.scheduler.retry.EndlessRetryStrategy;
 import io.camunda.zeebe.scheduler.retry.RetryStrategy;
 import io.camunda.zeebe.stream.api.EventFilter;
+import io.camunda.zeebe.util.VisibleForTesting;
 import io.camunda.zeebe.util.exception.UnrecoverableException;
 import io.camunda.zeebe.util.health.FailureListener;
 import io.camunda.zeebe.util.health.HealthMonitorable;
@@ -52,6 +52,8 @@ import org.slf4j.Logger;
 
 public final class ExporterDirector extends Actor implements HealthMonitorable, LogRecordAwaiter {
 
+  private static final String ERROR_MESSAGE_DESERIALIZATION_ERROR_EXPORTING_ABORTED =
+      "Expected to export record '{}' successfully, but exception was thrown when deserializing the record.";
   private static final String ERROR_MESSAGE_EXPORTING_ABORTED =
       "Expected to export record '{}' successfully, but exception was thrown.";
   private static final String ERROR_MESSAGE_RECOVER_FROM_SNAPSHOT_FAILED =
@@ -69,7 +71,6 @@ public final class ExporterDirector extends Actor implements HealthMonitorable, 
   private final ExporterMetrics metrics;
   private final String name;
   private final RetryStrategy exportingRetryStrategy;
-  private final RetryStrategy recordWrapStrategy;
   private final Set<FailureListener> listeners = new HashSet<>();
   private LogStreamReader logStreamReader;
   private EventFilter eventFilter;
@@ -96,6 +97,14 @@ public final class ExporterDirector extends Actor implements HealthMonitorable, 
 
   public ExporterDirector(
       final ExporterDirectorContext context, final ExporterPhase exporterPhase) {
+    this(context, exporterPhase, Function.identity());
+  }
+
+  @VisibleForTesting
+  ExporterDirector(
+      final ExporterDirectorContext context,
+      final ExporterPhase exporterPhase,
+      final Function<RecordExporter, RecordExporter> recorderExporter) {
     name = context.getName();
     logStream = Objects.requireNonNull(context.getLogStream());
     partitionId = logStream.getPartitionId();
@@ -114,9 +123,9 @@ public final class ExporterDirector extends Actor implements HealthMonitorable, 
             .collect(Collectors.toCollection(ArrayList::new));
     metrics = new ExporterMetrics(meterRegistry);
     metrics.initializeExporterState(exporterPhase);
-    recordExporter = new RecordExporter(metrics, containers, partitionId, clock);
+    recordExporter =
+        recorderExporter.apply(new RecordExporter(metrics, containers, partitionId, clock));
     exportingRetryStrategy = new BackOffRetryStrategy(actor, Duration.ofSeconds(10));
-    recordWrapStrategy = new EndlessRetryStrategy(actor);
     zeebeDb = context.getZeebeDb();
     this.exporterPhase = exporterPhase;
     partitionMessagingService = context.getPartitionMessagingService();
@@ -636,37 +645,29 @@ public final class ExporterDirector extends Actor implements HealthMonitorable, 
   }
 
   private void exportEvent(final LoggedEvent event) {
-    final ActorFuture<Boolean> wrapRetryFuture =
-        recordWrapStrategy.runWithRetry(
-            () -> {
-              recordExporter.wrap(event);
-              return true;
-            },
-            this::isClosed);
+    try {
+      recordExporter.wrap(event);
+    } catch (final Exception exception) {
+      LOG.warn(ERROR_MESSAGE_DESERIALIZATION_ERROR_EXPORTING_ABORTED, event, exception);
+      onFailure();
+      return;
+    }
+
+    final ActorFuture<Boolean> retryFuture =
+        exportingRetryStrategy.runWithRetry(recordExporter::export, this::isClosed);
 
     actor.runOnCompletion(
-        wrapRetryFuture,
-        (b, t) -> {
-          assert t == null : "Throwable must be null";
-
-          final ActorFuture<Boolean> retryFuture =
-              exportingRetryStrategy.runWithRetry(recordExporter::export, this::isClosed);
-
-          actor.runOnCompletion(
-              retryFuture,
-              (bool, throwable) -> {
-                if (throwable != null) {
-                  LOG.error(ERROR_MESSAGE_EXPORTING_ABORTED, event, throwable);
-                  onFailure();
-                } else {
-                  logStream
-                      .getFlowControl()
-                      .onExported(recordExporter.getTypedEvent().getPosition());
-                  metrics.eventExported(recordExporter.getTypedEvent().getValueType());
-                  inExportingPhase = false;
-                  actor.submit(this::readNextEvent);
-                }
-              });
+        retryFuture,
+        (bool, throwable) -> {
+          if (throwable != null) {
+            LOG.error(ERROR_MESSAGE_EXPORTING_ABORTED, event, throwable);
+            onFailure();
+          } else {
+            logStream.getFlowControl().onExported(recordExporter.getTypedEvent().getPosition());
+            metrics.eventExported(recordExporter.getTypedEvent().getValueType());
+            inExportingPhase = false;
+            actor.submit(this::readNextEvent);
+          }
         });
   }
 

--- a/zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirector.java
+++ b/zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirector.java
@@ -429,7 +429,10 @@ public final class ExporterDirector extends Actor implements HealthMonitorable, 
         failure,
         failure);
     actor.fail(failure);
+    updateHealthStatusWithError(failure);
+  }
 
+  private void updateHealthStatusWithError(final Throwable failure) {
     if (failure instanceof UnrecoverableException) {
       healthReport = HealthReport.dead(this).withIssue(failure, clock.instant());
 
@@ -649,6 +652,7 @@ public final class ExporterDirector extends Actor implements HealthMonitorable, 
       recordExporter.wrap(event);
     } catch (final Exception exception) {
       LOG.warn(ERROR_MESSAGE_DESERIALIZATION_ERROR_EXPORTING_ABORTED, event, exception);
+      updateHealthStatusWithError(new UnrecoverableException(exception));
       onFailure();
       return;
     }

--- a/zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/stream/RecordExporter.java
+++ b/zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/stream/RecordExporter.java
@@ -16,7 +16,7 @@ import io.camunda.zeebe.stream.impl.records.TypedRecordImpl;
 import java.time.InstantSource;
 import java.util.List;
 
-final class RecordExporter {
+class RecordExporter {
 
   private final RecordValues recordValues = new RecordValues();
   private final RecordMetadata rawMetadata = new RecordMetadata();

--- a/zeebe/broker/src/test/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirectorTest.java
+++ b/zeebe/broker/src/test/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirectorTest.java
@@ -37,6 +37,7 @@ import io.camunda.zeebe.protocol.record.intent.DeploymentIntent;
 import io.camunda.zeebe.protocol.record.intent.IncidentIntent;
 import io.camunda.zeebe.protocol.record.intent.JobIntent;
 import io.camunda.zeebe.stream.impl.SkipPositionsFilter;
+import io.camunda.zeebe.util.health.HealthStatus;
 import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -487,6 +488,7 @@ public final class ExporterDirectorTest {
 
   @Test
   public void shouldNotRetryExportingOnDeserializationException() {
+    // given
     final var recordExporterRef = new AtomicReference<RecordExporter>();
     rule.startExporterDirector(
         exporterDescriptors,
@@ -497,11 +499,21 @@ public final class ExporterDirectorTest {
     final var recordExporter = recordExporterRef.get();
     doThrow(new RuntimeException("deserialization exception")).when(recordExporter).wrap(any());
 
+    // when
     final long eventPosition1 = writeEvent();
 
+    // then
     verify(recordExporter, timeout(10000))
         .wrap(argThat(evt -> evt.getPosition() == eventPosition1));
     verifyNoMoreInteractions(recordExporter);
+    Awaitility.await("Until director is unhealthy")
+        .untilAsserted(
+            () ->
+                assertThat(rule.getDirector().getHealthReport().status())
+                    .isEqualTo(HealthStatus.DEAD));
+    assertThat(rule.getDirector().getPhase())
+        .succeedsWithin(Duration.ofSeconds(5))
+        .satisfies(phase -> assertThat(phase).isEqualTo(ExporterPhase.CLOSED));
   }
 
   @Test

--- a/zeebe/broker/src/test/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirectorTest.java
+++ b/zeebe/broker/src/test/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirectorTest.java
@@ -10,6 +10,7 @@ package io.camunda.zeebe.broker.exporter.stream;
 import static io.camunda.zeebe.test.util.TestUtil.doRepeatedly;
 import static io.camunda.zeebe.test.util.TestUtil.waitUntil;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.argThat;
 import static org.mockito.Mockito.any;
 import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.doThrow;
@@ -18,6 +19,7 @@ import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.timeout;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
 
 import io.camunda.zeebe.broker.exporter.repo.ExporterDescriptor;
 import io.camunda.zeebe.broker.exporter.util.ControlledTestExporter;
@@ -46,6 +48,7 @@ import java.util.Set;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import org.awaitility.Awaitility;
 import org.junit.Before;
@@ -483,6 +486,25 @@ public final class ExporterDirectorTest {
   }
 
   @Test
+  public void shouldNotRetryExportingOnDeserializationException() {
+    final var recordExporterRef = new AtomicReference<RecordExporter>();
+    rule.startExporterDirector(
+        exporterDescriptors,
+        ExporterPhase.EXPORTING,
+        recordExporter -> failingRecordExporter(recordExporter, recordExporterRef));
+    Awaitility.await("exporter director has started")
+        .untilAsserted(() -> assertThat(recordExporterRef.get()).isNotNull());
+    final var recordExporter = recordExporterRef.get();
+    doThrow(new RuntimeException("deserialization exception")).when(recordExporter).wrap(any());
+
+    final long eventPosition1 = writeEvent();
+
+    verify(recordExporter, timeout(10000))
+        .wrap(argThat(evt -> evt.getPosition() == eventPosition1));
+    verifyNoMoreInteractions(recordExporter);
+  }
+
+  @Test
   public void shouldExecuteScheduledTask() throws Exception {
     // given
     final CountDownLatch timerTriggerLatch = new CountDownLatch(1);
@@ -810,5 +832,10 @@ public final class ExporterDirectorTest {
     startExporterDirector(List.of(descriptor));
 
     return exporter;
+  }
+
+  private RecordExporter failingRecordExporter(
+      final RecordExporter recordExporter, final AtomicReference<RecordExporter> exporterRef) {
+    return exporterRef.updateAndGet(ignored -> spy(recordExporter));
   }
 }

--- a/zeebe/broker/src/test/java/io/camunda/zeebe/broker/exporter/stream/ExporterRule.java
+++ b/zeebe/broker/src/test/java/io/camunda/zeebe/broker/exporter/stream/ExporterRule.java
@@ -31,6 +31,7 @@ import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
 import java.time.Duration;
 import java.util.List;
 import java.util.Set;
+import java.util.function.Function;
 import java.util.stream.Collectors;
 import org.junit.rules.ExternalResource;
 import org.junit.rules.RuleChain;
@@ -107,6 +108,13 @@ public final class ExporterRule implements TestRule {
 
   public void startExporterDirector(
       final List<ExporterDescriptor> exporterDescriptors, final ExporterPhase phase) {
+    startExporterDirector(exporterDescriptors, phase, Function.identity());
+  }
+
+  public void startExporterDirector(
+      final List<ExporterDescriptor> exporterDescriptors,
+      final ExporterPhase phase,
+      final Function<RecordExporter, RecordExporter> recordExporter) {
     final var stream = streams.getLogStream(STREAM_NAME);
     final var runtimeFolder = streams.createRuntimeFolder(stream);
     capturedZeebeDb = spy(zeebeDbFactory.createDb(runtimeFolder.toFile()));
@@ -132,7 +140,7 @@ public final class ExporterRule implements TestRule {
             .meterRegistry(new SimpleMeterRegistry())
             .positionsToSkipFilter(positionsToSkipFilter);
 
-    director = new ExporterDirector(context, phase);
+    director = new ExporterDirector(context, phase, recordExporter);
     director.startAsync(actorSchedulerRule.get()).join();
   }
 

